go/cmd/betabot/main.go

```diff
@@ -476,18 +551,31 @@ func (bot *Bot) interactUserMessage(ctx context.Context, body string, conversati
 }
 
 func (bot *Bot) saveStore() {
-	bot.storeMutex.Lock()
-	defer bot.storeMutex.Unlock()
-
-	// marshal
-	data, err := json.MarshalIndent(bot.store, "", "  ")
-	if err != nil {
-		panic(fmt.Errorf("marshal: %w", err))
-	}
+	var done bool
+	// Prevent concurrent saves
+	bot.storeOnce.Do(func() {
```

Comment:
what's the advantage of this VS a simple mutex?

Reply:
> what's the advantage of this VS a simple mutex?

Multiple call to save at the same time are merged as one.

Reply:
that's a good idea, however, I think it's not working as expected:

* in a case, we can call it concurrently which may not guarantee that the object is the same, so we may just save the first call and lose the second one -> bad
* in the other case, we want to protect this by having a mutex to protect the base object, which makes this workflow useless

or I'm missing something?

Reply:
> in the other case, we want to protect this by having a mutex to protect the base object, which makes this workflow useless

That why it's called a last time at the closing of the program.

> In a case, we can call it concurrently which may not guarantee that the object is the same, so we may just save the first call and lose the second one -> bad

True, I was thinking this wasn't an issue because at worst this can't propagate very far (values are not lost they are not written to disk, but the next true save will write them).
I've fixed that by releasing the locks after the reset. (For this situation to happen again you would need to call save while editing the store (wich thx to some mutex is impossible and would just deadlocks)).